main.py

```diff
@@ -0,0 +1,267 @@
+        auto_inline_consumers(sch, b_spatial)
+        sch.annotate(bx, ann_key="pragma_auto_unroll_max_step", ann_val=unroll_depth)
+        sch.annotate(bx, ann_key="pragma_unroll_explicit", ann_val=1)
+
+        return sch
+
+
+class Softmax(ScheduleRule):
+    def __init__(self):
+        ...
+
+    def _initialize_with_tune_context(self, context: ms.TuneContext) -> None:
+        pass
+
+    def clone(self) -> ScheduleRule:
+        return Softmax()
+
+    def apply(self, mod: IRModule) -> Optional[tir.Schedule]:  # pylint: disable=too-many-locals
+        sch = tir.Schedule(mod)
+        try:
+            b_reduce_0 = sch.get_block("T_softmax_maxelem")
+            b_reduce_1 = sch.get_block("T_softmax_expsum")
+            b_spatial = sch.get_block("T_softmax_norm")
+        except:  # pylint: disable=bare-except
+            return None
+
+        len_tx: int = 256
+        unroll_depth: int = 256
+
+        sch.compute_inline(sch.get_producers(b_reduce_1)[0])
+
+        loops = sch.get_loops(b_spatial)
+        bx = sch.fuse(*loops[:-1])  # pylint: disable=invalid-name
+        _, tx = sch.split(loops[-1], [None, len_tx])  # pylint: disable=invalid-name
+        sch.bind(bx, "blockIdx.x")
+        sch.bind(tx, "threadIdx.x")
+
+        sch.set_scope(b_reduce_1, buffer_index=0, storage_scope="shared")
+        sch.compute_at(b_reduce_1, bx, preserve_unit_loops=True)
+        _, tx = sch.split(  # pylint: disable=invalid-name
+            sch.get_loops(b_reduce_1)[-1],
+            [None, len_tx],
+        )
+        sch.bind(tx, "threadIdx.x")
+
+        sch.set_scope(b_reduce_0, buffer_index=0, storage_scope="shared")
+        sch.compute_at(b_reduce_0, bx, preserve_unit_loops=True)
+        _, tx = sch.split(  # pylint: disable=invalid-name
+            sch.get_loops(b_reduce_0)[-1],
+            [None, len_tx],
+        )
+        sch.bind(tx, "threadIdx.x")
+
+        sch.annotate(bx, ann_key="pragma_auto_unroll_max_step", ann_val=unroll_depth)
+        sch.annotate(bx, ann_key="pragma_unroll_explicit", ann_val=1)
+
+        auto_inline_consumers(sch, b_spatial)
+        return [sch]
+
+
+def main():
+    from tvm.dlight.testing import mod_decode, mod_decode_gemv, mod_norm, mod_softmax
+
+    gen = ScheduleGenerator(
+        rules=[
+            DecodeGemv(),
+            Decode(),
+            Normalization(),
+            Softmax(),
+        ]
+    )
+
+    for py_mod in [
+        mod_decode,
+        mod_decode_gemv,
+        mod_norm,
+        mod_softmax,  # Needs to upstream `compute-inline`
+    ]:
+        i = 1
+        while True:
+            try:
+                func = py_mod.Module[f"func{i}"]
+            except:  # pylint: disable=bare-except
+                break
+            else:
+                print(f"Working on {py_mod}::func{i}")
+            i += 1
+            mod = IRModule.from_expr(func.with_attr("global_symbol", "main"))
+            schedules = gen.generate_design_space(mod)
```

Comment:
Let us structure the testcase UT same as Before/After test, e.g. one softmax function before and after with structural equality test.

Atm we only test that the schedule exists